### ISSUE TITLE
[heft] Jest build stage - Failed to read cache file 

### ISSUE THREAD

**Issue by josh-biddick** [microsoft/rushstack#2800]
## Summary

Often when running `heft test` in a terminal, we get the following error when Jest test are running:

![image](https://user-images.githubusercontent.com/74222743/124990574-a9979f80-e094-11eb-96d8-e962245ed739.png)

We don't get this error when running the tests via GitHub Actions. It seems to only present itself when there is another local process locking file permissions. In our particular case, we run GitKraken. Closing this and re-running the tests seems to resolve the issue.

## Repro steps

1. Run GitKraken and open the repo that contains the tests
2. Run `heft test --clean`
3. Wait for the tests to run and reproduce the above error

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@rushstack/heft` version? | 0.34.3 |
| Operating system? | Windows 10  |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 12.16.1 |

**Comment by octogonz** [comment]
Could this be related to https://github.com/microsoft/rushstack/issues/2712?  In that case a subprocess from a previous command is still running when `heft test` starts.  The new Jest process gets confused because a file gets deleted while it is trying to read it. The deleting is caused by the previous process cleaning up the folder.

I'm not sure if that applies here.  But the callstack from your screenshot is definitely not atomic:

```
const readCacheFile = cachePath => {
  if (!fs().existsSync(cachePath)) { 👈👈👈
    return null;
  }

  let fileData;

  try {
    fileData = fs().readFileSync(cachePath, 'utf8'); 👈👈👈 assumes nobody else is modifying this folder
  } catch (e) {
    e.message =
      'jest: failed to read cache file: ' +
      cachePath +
      '\nFailure message: ' +
      e.message;
    removeFile(cachePath);
    throw e;
  }

  if (fileData == null) {
    // We must have somehow created the file but failed to write to it,
    // let's delete it and retry.
    removeFile(cachePath);
  }

  return fileData;
};
```

**Comment by octogonz** [comment]
Is there a way to tell GitKraken to leave certain folders alone?  Seems like its crawler is not honoring `.gitignore`.

It would be very helpful to run your repo with [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) and then apply a filter to see which processes touched that file path and what they were doing.

**Comment by octogonz** [comment]
This feels like a GitKraken bug or a Jest bug, but let's keep this issue open until we rule out the possibility that Heft could implement a workaround for it.

**Comment by josh-biddick** [comment]
@octogonz Thanks. I'll keep an eye on it over the next few days and try to capture the events with Process Monitor as suggested.

**Comment by iclanton** [comment]
@josh-biddick, were you able to learn anything more about this issue?

**Comment by josh-biddick** [comment]
@iclanton No, although I have got Process Monitor installed for when/if it appears again. Feel free to close if needed.
